### PR TITLE
Adds support for RGB colored text for chat

### PR DESCRIPTION
**Pull request by ag3killstreak**
Uses color character with 3 additional ascii characters in range of 0x20 to 0x7E to output RGB values to the client text.

Example:
ClientPrint(null, 3, "\x08~!!RED\x08~A!ORANGE\x08~~!YELLOW\x08!~!GREEN\x08!~~AQUA\x08!A~LIGHT_BLUE\x08!!~BLUE\x08~A`PINK\x08~a~PINK");

This is further simplified by the TextColor vscript function that accepts RGB and outputs the closest ascii range.

Example:

ClientPrint(null, 3, TextColor(255, 0 0) +"RED" + " " + TextColor(163, 73, 164)+ "PURPLE" + " " + TextColor(255, 255, 255)+ "WHITE" + " " + TextColor(0, 0, 0)+ "BLACK" + " " + TextColor(155, 155, 155)+ "SILVER");

### CODE REVIEW

File: src/game/server/vscript_server.cpp [hunk 1]
```diff
@@ -1201,6 +1200,57 @@ static void Script_Say( HSCRIPT hPlayer, const char *pText )
...
+	char R_Char = (char)(32 + (outputMod_R * 94));
+	char G_Char = (char)(32 + (outputMod_G * 94));
+	char B_Char = (char)(32 + (outputMod_B * 94));
+
+
+	//Format a string with results as 4 characters.
+	char outputChars[5];
+	Q_snprintf(outputChars, sizeof(outputChars), "%c%c%c%c", Con_Char, R_Char, G_Char, B_Char);
+
+	return outputChars;
```
Comment (geotavros) [inline src/game/server/vscript_server.cpp:1252]: Return of a pointer to the out-of-scope variable outputChars.

Comment (ag3killstreak) [reply src/game/server/vscript_server.cpp:1252]: Are you suggesting something like the following?

	char outputChars[5];
	Q_snprintf(outputChars, sizeof(outputChars), "%c%c%c%c", Con_Char, R_Char, G_Char, B_Char);

	return ScriptVariant_t(outputChars, true);

**Comment by wawawawawawawa** [comment]
`COLOR_YELLOW = (1).tochar();
COLOR_BLUE = (5).tochar();
COLOR_LIGHTBLUE = (3).tochar();
COLOR_GREEN = (4).tochar();
COLOR_PURPLE = (6).tochar();
COLOR_RED = (7).tochar();

ClientPrint(null, 3, COLOR_BLUE + "I'm blue! " + COLOR_RED + "I'm red!" + COLOR_BLUE + "and I'm back to blue!");`

This allows you to use colored text with clientprint

**Comment by ag3killstreak** [comment]
> ```
> COLOR_YELLOW = (1).tochar();
> COLOR_BLUE = (5).tochar();
> COLOR_LIGHTBLUE = (3).tochar();
> COLOR_GREEN = (4).tochar();
> COLOR_PURPLE = (6).tochar();
> COLOR_RED = (7).tochar();
> 
> ClientPrint(null, 3, COLOR_BLUE + "I'm blue! " + COLOR_RED + "I'm red!" + COLOR_BLUE + "and I'm back to blue!");
> ```
> 
> This allows you to use colored text with clientprint (credits @jhh8)

My code allows around 830,584 possible different colors. The problem with your quoted method is only so many characters can be used as color characters. Once they're gone, They're gone.

**Comment by jhh8** [comment]
fyi using buffer size more than 512 bytes doesnt work as theres a limit of bytes that can be send in a usermessage. 
DLL_MessageEnd:  Refusing to send user message TextMsg of 512 bytes to client, user message size limit is 511 bytes

**Comment by ag3killstreak** [comment]
> fyi using buffer size more than 512 bytes doesnt work as theres a limit of bytes that can be send in a usermessage. DLL_MessageEnd: Refusing to send user message TextMsg of 512 bytes to client, user message size limit is 511 bytes

I did know there was a limit. I however did not know it was 512. That is useful information, thank you.

File: src/game/client/hud_basechat.h [hunk 2]
```diff
@@ -131,6 +132,7 @@ class CBaseHudChatLine : public vgui::RichText
 	void			SetNameColor( Color cColor ){ m_clrNameColor = cColor; 	}
 		
 	virtual void	PerformFadeout( void );
+	int				CBaseHudChatLine::TranslateChannelRange(byte inputval);
```
Comment (BenLubar) [inline src/game/client/hud_basechat.h:135]:
```suggestion
	static int		TranslateChannelRange(byte inputval);
```

File: src/game/client/swarm/vgui/asw_hud_chat.cpp [hunk 3]
```diff
@@ -198,8 +198,8 @@ void CHudChat::MsgFunc_TextMsg( bf_read &msg )
 	char szString[2048];
 	int msg_dest = msg.ReadByte();
 
-	wchar_t szBuf[5][128];
-	wchar_t outputBuf[256];
+	wchar_t szBuf[5][512]; //Changed to 512 to increase buffer size for custom color.
+	wchar_t outputBuf[512]; //Changed to 512 to increase buffer size for custom color.
```
Comment (BenLubar) [inline src/game/client/swarm/vgui/asw_hud_chat.cpp:202]:
```suggestion
	wchar_t szBuf[5][512];
	wchar_t outputBuf[512];
```
We've got public revision history and we're not expecting any upstream (Valve Alien Swarm) changes; we don't need comments describing changes.

File: src/game/server/vscript_server.cpp [hunk 4]
```diff
@@ -1485,6 +1536,7 @@ bool VScriptServerInit()
 				ScriptRegisterFunction( g_pScriptVM, CreateProp, "Create a physics prop" );
 				ScriptRegisterFunctionNamed( g_pScriptVM, Script_Say, "Say", "Have player say string" );
 				ScriptRegisterFunctionNamed( g_pScriptVM, Script_ClientPrint, "ClientPrint", "Print a client message" );
+				ScriptRegisterFunctionNamed(g_pScriptVM,  Script_TextColor, "TextColor", "Gets the translated ASCII characters for an RBG input.");
```
Comment (BenLubar) [inline src/game/server/vscript_server.cpp:1539]:
```suggestion
				ScriptRegisterFunctionNamed( g_pScriptVM, Script_TextColor, "TextColor", "Gets the translated ASCII characters for an RGB input." );
```

File: src/game/server/vscript_server.cpp [hunk 5]
```diff
@@ -1201,6 +1201,57 @@ static void Script_Say( HSCRIPT hPlayer, const char *pText )
...
+	char Con_Char = (char)0x08;
+	//pass float modifiers multiplied by max ASCII translation base then increment by 32 which is float ASCII offset
+	char R_Char = (char)(32 + (outputMod_R * 94));
+	char G_Char = (char)(32 + (outputMod_G * 94));
+	char B_Char = (char)(32 + (outputMod_B * 94));
+
+
+	//Format a string with results as 4 characters.
+	char outputChars[5];
+	Q_snprintf(outputChars, sizeof(outputChars), "%c%c%c%c", Con_Char, R_Char, G_Char, B_Char);
```
Comment (BenLubar) [inline src/game/server/vscript_server.cpp:1250]:
```suggestion
	char outputChars[5]{};
	outputChars[0] = '\x08'; // COLOR_INPUTCUSTOMCOL
	// pass float modifiers multiplied by max ASCII translation base then increment by 32 which is float ASCII offset
	outputChars[1] = (char)( 32 + ( outputMod_R * 94 ) );
	outputChars[2] = (char)( 32 + ( outputMod_G * 94 ) );
	outputChars[3] = (char)( 32 + ( outputMod_B * 94 ) );
```

File: src/game/client/hud_basechat.cpp [hunk 6]
```diff
@@ -1758,6 +1817,13 @@ void CBaseHudChat::ChatPrintf( int iPlayerIndex, int iFilter, const char *fmt, .
...
 	while ( *pmsg && ( *pmsg == '\n' || ( *pmsg > 0 && *pmsg < COLOR_MAX ) ) )
 	{
+		//This is code I added to clear the 3 bytes that come after the custom RGB character.
+		if (*pmsg == COLOR_INPUTCUSTOMCOL)
+		{
+			pmsg++;
+			pmsg++;
+			pmsg++;
+		}
 		pmsg++;
```
Comment (BenLubar) [inline src/game/client/hud_basechat.cpp:1827]:
```suggestion
		if ( *pmsg == COLOR_INPUTCUSTOMCOL )
			pmsg += 4;
		else
			pmsg++;
```

File: src/game/client/hud_basechat.cpp [hunk 7]
```diff
@@ -1535,7 +1560,41 @@ void CBaseHudChatLine::InsertAndColorizeText( wchar_t *buf, int clientIndex )
...
+				{
+					m_textRanges[count - 1].end = range.start - 3; //3 Offset accounts for color channel translation characters
+				}
+
+				m_textRanges.AddToTail(range);
+				//Pass by all color channel characters
+				++txt;
+				++txt;
+				++txt;
+				++txt;
```
Comment (BenLubar) [inline src/game/client/hud_basechat.cpp:1595]:
```suggestion
				txt += 4;
```

**Review by BenLubar — DISMISSED** [review]
some code cleanliness suggestions

**Comment by BenLubar** [comment]
@ag3killstreak what's your Steam profile? (I want to give you a Programmer medal.)

**Comment by ag3killstreak** [comment]
> @ag3killstreak what's your Steam profile? (I want to give you a Programmer medal.)

https://steamcommunity.com/id/4AG3154Be/

**Pull request merged**